src/Compilers/Core/AnalyzerDriver/AnalyzerManager.cs

```diff
@@ -61,7 +61,41 @@ internal class AnalyzerManager
                 }, cancellationToken));
         }
 
-        private Task<HostSessionStartAnalysisScope> GetSessionAnalysisScopeAsync(
+        private async Task<HostCompilationStartAnalysisScope> GetCompilationAnalysisScopeAsync(
+            DiagnosticAnalyzer analyzer,
+            HostSessionStartAnalysisScope sessionScope,
+            Compilation compilation,
+            Action<Diagnostic> addDiagnostic,
+            AnalyzerOptions analyzerOptions,
+            Func<Exception, DiagnosticAnalyzer, bool> continueOnAnalyzerException,
+            CancellationToken cancellationToken)
+        {
+            try
+            {
+                return await GetCompilationAnalysisScopeCoreAsync(analyzer, sessionScope,
+                    compilation, addDiagnostic, analyzerOptions, continueOnAnalyzerException, cancellationToken).ConfigureAwait(false);
+            }
+            catch (TaskCanceledException)
+            {
+                // Task to compute the scope was cancelled.
+                // Clear the entry in scope map for analyzer, so we can attempt a retry.
+                var compilationActionsMap = _compilationScopeMap.GetOrCreateValue(compilation);
+                Task<HostCompilationStartAnalysisScope> cancelledTask;
+                compilationActionsMap.TryRemove(analyzer, out cancelledTask);
+
+                // Don't attempt retry if our cancellation token fired as the retry will also fail.
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return new HostCompilationStartAnalysisScope(sessionScope);
```

Comment:
I'll replace the if statement with cancellationToken.ThrowIfCancellationRequested.